### PR TITLE
Add media

### PR DESCRIPTION
**Pull request by brinxmat**

### CODE REVIEW

File: documentation/MediaInternet.json [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    "reference" : {
+      "type" : "Reference",
+      "publicationContext" : {
+        "type" : "Media",
+        "place" : "Elsewhere",
+        "programme" : "Norbert and Friends"
+      },
+      "doi" : "https://123123/213123.com",
+      "publicationInstance" : {
+        "type" : "MediaInternet",
```
Comment (elkebab) [inline documentation/MediaInternet.json:78]: bilde

I skissene ser det ut som at det er tenkt annerledes, hvor `publicationInstance.type` er en av: 
- Dokumenter
- Intervju
- Programledelse
- Programdeltagelse

Mens `Mediumtype` som virker å være det samme som `publicationInstance.contentType` for bl.a. artikkel i tidsskrift og kapittel

File: nva-datamodel-java/src/main/java/no/unit/nva/model/contexttypes/Media.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.Objects;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+public class Media implements PublicationContext {
+    public static final String PROGRAMME = "programme";
+    public static final String PLACE = "place";
+
+    @JsonProperty(PLACE)
+    private final String place;
+    @JsonProperty(PROGRAMME)
```
Comment (axthosarouris) [inline nva-datamodel-java/src/main/java/no/unit/nva/model/contexttypes/Media.java:17]: The  "programme" here is in a way connected to the associated publication instance.

File: nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/media/AbstractMediaInstance.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+public abstract class AbstractMediaInstance implements PublicationInstance<NullPages> {
+
+    private static final String PAGES = "pages";
+    public static final String CONTRIBUTION_TYPE = "contributionType";
+
+    @JsonProperty(CONTRIBUTION_TYPE)
+    private final ContributionType contributionType;
+    @JsonProperty(PAGES)
+    private static final NullPages pages = new NullPages();
```
Comment (axthosarouris) [inline nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/media/AbstractMediaInstance.java:20]: if many classes have this "NullPages" doesn't this mean that we have may two or more kinds of PublicationInstances? Those with Pages (printed?) and those without.

File: nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/media/AbstractMediaInstance.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    @JsonGetter(CONTRIBUTION_TYPE)
+    public ContributionType getContributionType() {
+        return contributionType;
+    }
+
+    @JsonGetter(PAGES)
+    @Override
+    public NullPages getPages() {
+        return pages;
```
Comment (axthosarouris) [inline nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/media/AbstractMediaInstance.java:34]: return "NullPages.instance()" where instance() returns the one and only NullPages object (and the constructor of NullPages is private);

Comment (axthosarouris) [reply nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/media/AbstractMediaInstance.java:34]: You can deprecate the NullPages constructor for now :)

File: nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/media/AbstractMediaInstance.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @JsonSetter(PAGES)
+    @Override
+    public abstract void setPages(NullPages pages);
```
Comment (axthosarouris) [inline nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/media/AbstractMediaInstance.java:39]: this should be the exact opposite of abstract. it should be final. The decision of what one can do here has been set by the type of the input parameter

File: nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/media/MediaProgrammeParticipation.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+public class MediaProgrammeParticipation extends AbstractMediaInstance {
+    @JsonCreator
+    public MediaProgrammeParticipation(@JsonProperty(CONTRIBUTION_TYPE) ContributionType contributionType) {
+        super(contributionType);
+    }
+
+    @Override
+    public void setPages(NullPages pages) {
+
```
Comment (axthosarouris) [inline nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/media/MediaProgrammeParticipation.java:17]: 4 times setPages(NullPages) -> make a final method in the abstract class

File: nva-datamodel-java/src/test/java/no/unit/nva/PublicationMapperTest.java [hunk 7]
```diff
@@ -81,6 +81,13 @@ public void canMapPublicationAndContextToPublicationResponse() throws Exception
         assertNotNull(response);
         assertThat(response.getId(), notNullValue());
         assertThatIdIsPresent(response);
+        assertThatTypIsPresent(response);
+    }
+
+    private void assertThatTypIsPresent(PublicationResponse response) throws JsonProcessingException {
+        String string = dataModelObjectMapper.writeValueAsString(response);
+        ObjectNode json = (ObjectNode) dataModelObjectMapper.readTree(string);
+        assertThat(json.has("type"), is(equalTo(true)));
```
Comment (axthosarouris) [inline nva-datamodel-java/src/test/java/no/unit/nva/PublicationMapperTest.java:90]: we can avoid the generic assertThat(assertMyExpectation(),is(true)) here by using this:
https://github.com/spotify/java-hamcrest#json-matchers

File: nva-datamodel-java/src/test/java/no/unit/nva/model/util/PublicationGenerator.java [hunk 8]
```diff
@@ -334,6 +347,22 @@ private static PublicationContext getPublishingContextArtisticDesign() {
         return new Lecture();
     }
 
+    private static PublicationInstance<NullPages> getProgrammeParticipation() {
```
Comment (axthosarouris) [inline nva-datamodel-java/src/test/java/no/unit/nva/model/util/PublicationGenerator.java:350]: It is time to create a RefenceGenerator and put this method in its own class.

File: nva-datamodel-java/src/test/java/no/unit/nva/model/util/PublicationGenerator.java [hunk 9]
```diff
@@ -349,57 +378,20 @@ private static PublicationContext getPublishingContextEvent() {
...
-            .withNpiSubjectHeading("Soulfulness")
-            .build();
-
-        return generatePublication(SortableIdentifier.next(), UUID.randomUUID(), Instant.now(), entityDescription);
-    }
 
     public static List<ResearchProject> getProjects() {
         return List.of(new ResearchProject.Builder()
-                           .withId(URI.create("http://link.to.cristin.example.org/123"))
+                           .withId(URI.create("https://link.to.cristin.example.org/123"))
```
Comment (axthosarouris) [inline nva-datamodel-java/src/test/java/no/unit/nva/model/util/PublicationGenerator.java:394]: comment: You can use randomUri() from testutils

**Review by axthosarouris — CHANGES_REQUESTED** [review]

**Pull request closed without merging**